plutus-tx/test/Plugin/ReadValue.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE DataKinds        #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS -fplugin Language.PlutusTx.Plugin -fplugin-opt Language.PlutusTx.Plugin:defer-errors -fplugin-opt Language.PlutusTx.Plugin:strip-context #-}
+-- the simplifier messes with things otherwise
+{-# OPTIONS_GHC   -O0 #-}
+{-# OPTIONS_GHC   -Wno-orphans #-}
+{-# OPTIONS_GHC   -fmax-simplifier-iterations=0 #-}  -- being paranoid
+-- this adds source notes which helps the plugin give better errors
+{-# OPTIONS_GHC   -g #-}
+
+module Plugin.ReadValue
+    ( readDyns
+    ) where
+
+import           Common
+import           PlcTestUtils
+
+import qualified Language.PlutusTx.Builtins                 as Builtins
+import           Language.PlutusTx.Lift
+import           Language.PlutusTx.Plugin
+
+import qualified Language.PlutusCore                        as PLC
+import qualified Language.PlutusCore.Constant               as PLC
+import qualified Language.PlutusCore.Constant.Dynamic       as PLC
+import qualified Language.PlutusCore.Evaluation.Result      as PLC
+import qualified Language.PlutusCore.Interpreter.CekMachine as PLC
+
+import           Data.ByteString.Lazy
+import           Data.Text.Prettyprint.Doc
+import           GHC.Generics
+import           Test.Tasty
+import           Test.Tasty.HUnit
+
+-- this module does lots of weird stuff deliberately
+{-# ANN module ("HLint: ignore"::String) #-}
+
+newtype MyNewtype = MyNewtype Int
+    deriving (Show, Eq)
+
+data MyMonoRecord = MyMonoRecord { mrA :: Int , mrB :: Int}
+    deriving (Show, Eq)
+
+data MyMonoData = Mono1 Int Int | Mono2 Int | Mono3 Int
+    deriving (Show, Eq)
+
+readCompiledCode :: PLC.KnownDynamicBuiltinType a => (a -> b) -> CompiledCode b -> b
```

Comment:
A comment would be good here - I was very confused for a bit as to why we were reading at the wrong type and then injecting it, but it's because we read datatypes as products.

Reply:
Done.